sandbox/modules/custom-codecs/src/main/java/org/opensearch/index/codec/customcodec/ZstdCompressionMode.java

```diff
@@ -0,0 +1,200 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.index.codec.customcodec;
+
+import com.github.luben.zstd.Zstd;
+import com.github.luben.zstd.ZstdCompressCtx;
+import com.github.luben.zstd.ZstdDecompressCtx;
+import com.github.luben.zstd.ZstdDictCompress;
+import com.github.luben.zstd.ZstdDictDecompress;
+import java.io.IOException;
+import org.apache.lucene.codecs.compressing.CompressionMode;
+import org.apache.lucene.codecs.compressing.Compressor;
+import org.apache.lucene.codecs.compressing.Decompressor;
+import org.apache.lucene.store.DataInput;
+import org.apache.lucene.store.DataOutput;
+import org.apache.lucene.store.ByteBuffersDataInput;
+import org.apache.lucene.util.ArrayUtil;
+import org.apache.lucene.util.BytesRef;
+
+/** Zstandard Compression Mode */
+public class ZstdCompressionMode extends CompressionMode {
+
+    private static final int NUM_SUB_BLOCKS = 10;
+    private static final int DICT_SIZE_FACTOR = 6;
+    private static final int DEFAULT_COMPRESSION_LEVEL = 6;
+
+    private final int compressionLevel;
+
+    /** default constructor */
+    protected ZstdCompressionMode() {
+        this.compressionLevel = DEFAULT_COMPRESSION_LEVEL;
+    }
+
+    /**
+     * Creates a new instance.
+     *
+     * @param compressionLevel The compression level to use.
+     */
+    protected ZstdCompressionMode(int compressionLevel) {
+        this.compressionLevel = compressionLevel;
+    }
+
+    @Override
+    public Compressor newCompressor() {
+        return new ZSTDCompressor(compressionLevel);
+    }
+
+    @Override
+    public Decompressor newDecompressor() {
+        return new ZSTDDecompressor();
+    }
+
+    /** zstandard compressor */
+    private static final class ZSTDCompressor extends Compressor {
+
+        private final int compressionLevel;
+        private byte[] compressedBuffer;
+
+        /** compressor with a given compresion level */
+        public ZSTDCompressor(int compressionLevel) {
+            this.compressionLevel = compressionLevel;
```

Comment:
This can be configured externally, right? Let's guard against invalid values (negative, very large, etc.)

Reply:
No, it cannot. Currently, there is no good way to handle additional configuration parameters for compressors besides the `index.codec` parameter which is used to set the codec to use. Same is true for the existing `BEST_COMPRESSION` codec; we cannot choose a level for it -- it's internally set in [Lucene at level 6](https://github.com/apache/lucene/blob/main/lucene/core/src/java/org/apache/lucene/codecs/lucene90/DeflateWithPresetDictCompressionMode.java#L55). May be this is something we need to handle later as it requires introducing a new configuration parameter and a patch to Lucene (at least to allow users to choose a compression level for deflate).

For the time being, I'll make the non-default c'tor `private`.